subxt/src/events/filter_events.rs

```diff
@@ -0,0 +1,418 @@
+// Copyright 2019-2022 Parity Technologies (UK) Ltd.
+// This file is part of subxt.
+//
+// subxt is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// subxt is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with subxt.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Filtering individual events from subscriptions.
+
+use super::Events;
+use crate::{
+    BasicError,
+    Config,
+    Event,
+    Phase,
+};
+use codec::Decode;
+use futures::{
+    Stream,
+    StreamExt,
+};
+use std::{
+    marker::Unpin,
+    task::Poll,
+};
+
+/// A stream which filters events based on the `Filter` param provided.
+/// If `Filter` is a 1-tuple of a single `Event` type, it will return every
+/// instance of that event as it's found. If `filter` is ` tuple of multiple
+/// `Event` types, it will return a corresponding tuple of `Option`s, where
+/// exactly one of these will be `Some(event)` each iteration.
+pub struct FilterEvents<'a, Sub: 'a, T: Config, Filter: EventFilter> {
+    // A subscription; in order for the Stream impl to apply, this will
+    // impl `Stream<Item = Result<Events<'a, T, Evs>, BasicError>> + Unpin + 'a`.
+    sub: Sub,
+    // Each time we get Events from our subscription, they are stored here
+    // and iterated through in future stream iterations until exhausted.
+    events: Option<
+        Box<
+            dyn Iterator<
+                    Item = Result<
+                        FilteredEventDetails<T::Hash, Filter::ReturnType>,
+                        BasicError,
+                    >,
+                > + 'a,
+        >,
+    >,
```

Comment:
This made me giggle, you can really tell `cargo fmt` was cussing and sweating here.